### ISSUE TITLE
session expired when some panel query failed

### ISSUE THREAD

**Issue by xaochuan1987** [grafana/grafana#16322]
**What happened**:
session expired when some panel query failed. I haven't met this when I used grafana 4.X.

**What you expected to happen**:
do not login again

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
After query failed, the browser like this:
![image](https://user-images.githubusercontent.com/23429939/55310738-cc898100-5493-11e9-861e-6659d3a3cb1a.png)

grafana query log:
t=2019-04-01T15:21:40+0800 lvl=info msg="Request Completed" logger=context userId=3 orgId=1 uname=XXX method=POST path=/api/datasources/proxy/25/_msearch status=502 remote_addr=172.18.83.128 time_ms=57990 size=0 referer=XXX

**Environment**:
- Grafana version: 6.0.1
- Data source type & version: elasticsearch 5.x
- OS Grafana is installed on: Centos6
- User OS & Browser: Mac/Chrome
- Grafana plugins: 
- Others:

**Comment by torkelo** [comment]
What authentication for elasticsearch did you use?

**Comment by xaochuan1987** [comment]
> What authentication for elasticsearch did you use?

My elasticsearch has no authentication, every client can access data by http.

**Comment by stuartnelson3** [comment]
We are also experiencing this when using v6.1.0.

Environment:

    Grafana version: 6.1.0
    Data source type & version: prometheus 2.6
    OS Grafana is installed on: Debian
    User OS & Browser: Linux/Firefox (reported also on macos/{firefox,chrome})
    Grafana plugins: grafana-clock-panel grafana-piechart-panel grafana-simple-json-datasource raintank-worldping-app mtanda-histogram-panel jdbranham-diagram-panel camptocamp-prometheus-alertmanager-datasource
    Others:

When requests fail, for example if I lose connection to our VPN, it appears that when I reload the page after reconnecting to the VPN, I've been logged out.

**Comment by torkelo** [comment]
Not sure what could be the cause. Something with the VPN and clearing your cookies? 

any ideas @marefr ?

**Comment by stuartnelson3** [comment]
On a dashboard with many panels, a transient failure in any of them when refreshing also logs the user out.

**Comment by marefr** [comment]
@stuartnelson3 would like the response details from chrome developer tools when you receive the Unauthorized message. Use the network tab and check the **Preserve log** in case you're being redirected after Unauthorized message. I would expect the `/api/datasources/proxy/25/_msearch` to return a 401 response status.

**Comment by stuartnelson3** [comment]
I haven't been receiving an unauthorized message, that was the reporter of this issue.

**Comment by marefr** [comment]
@stuartnelson3 sorry for that. But since you explain "a transient failure in any of them when refreshing also logs the user out" I would like to see the request/response details this transient failure and chrome developer tools (Use the network tab and check the Preserve log in case you're being redirected after Unauthorized message). Thanks

**Comment by marefr** [comment]
@xaochuan1987 would like the response details from chrome developer tools when you receive the Unauthorized message. Use the network tab and check the **Preserve log** in case you're being redirected after Unauthorized message. I would expect the `/api/datasources/proxy/25/_msearch` to return a 401 response status.

**Comment by stuartnelson3** [comment]
Failing request headers:
```
GET /api/datasources/proxy/501/api/v1/query_range?<prometheus-query>
HTTP/1.1
Host: grafana.int.s-cloud.net
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
X-Grafana-Org-Id: 1
Connection: keep-alive
Referer: http://<url>/d/qe1BekPmz/dashboard?orgId=1&from=now-3h&to=now&refresh=10s
Cookie: grafana_user=<email>; grafana_remember=<cookie>
Pragma: no-cache
Cache-Control: no-cache
```
There is no response. In this case, the request failed because I lost connection to the VPN.

When I re-establish connection, requests are then denied (because something caused me to log out in the failure it seems):
request headers
```
POST /api/dashboards/db/ HTTP/1.1
Host: <url>
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
X-Grafana-Org-Id: 1
Content-Type: application/json;charset=utf-8
Content-Length: 10616
Connection: keep-alive
Referer: http://<url>/d/qe1BekPmz/dashboard?orgId=1&from=now-3h&to=now&refresh=10s
Cookie: grafana_user=<email>; grafana_remember=<cookie>
Pragma: no-cache
Cache-Control: no-cache
```
response headers
```
HTTP/1.1 403 Forbidden
Content-Length: 45
Content-Type: application/json
Date: Tue, 09 Apr 2019 07:49:31 GMT
```
response
```
{"message":"Access denied to save dashboard"}
```

To reproduce:

1. Load a dashboard
2. Cause a panel request to fail (lose internet connection, e.g.)
3. Re-establish connection
4. UI still shows me as logged in, but any authenticated actions (such as saving, see above) are now unauthorized
5. Reloading the page shows that I am not logged in

**Comment by marefr** [comment]
@stuartnelson3 are you sure you're using v6.1.0? You should have a cookie named `grafana_session` since Grafana v6.0 when logged in.

**Comment by stuartnelson3** [comment]
I do have that cookie in local storage, but it is not included in the headers.

from the running binary:
```
% grafana-server -v
Version 6.1.0 (commit: 03346b6, branch: HEAD)
```

**Comment by marefr** [comment]
> I do have that cookie in local storage, but it is not included in the headers.

Not sure what you mean by that. Cookies should not be in local storage?

What [root_url](http://docs.grafana.org/installation/configuration/#root-url) have you configured in Grafana and what url are you using for accessing Grafana? Are you accessing Grafana embedded in an iframe?

**Comment by stuartnelson3** [comment]
>Not sure what you mean by that. Cookies should not be in local storage?

it's in cookies, then. it's stored locally on my machine:

![image](https://user-images.githubusercontent.com/1398104/55799014-a9cd1d00-5ad0-11e9-8a69-72877ccf95ce.png)

after being forced to log in again, it is available in the headers on a request:

```
GET /api/search HTTP/1.1
Host: <host>
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:67.0) Gecko/20100101 Firefox/67.0
Accept: application/json, text/plain, */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
X-Grafana-Org-Id: 1
Connection: keep-alive
Referer: <host>
Cookie: grafana_user=<user>; grafana_remember=<remember>; grafana_session=<session>
Pragma: no-cache
Cache-Control: no-cache
```

>What root_url have you configured in Grafana and what url are you using for accessing Grafana? Are you accessing Grafana embedded in an iframe?

`server_domain` is configured to the same URL as what we are using to access grafana. It is not being access through an iframe. google oauth authentication is successful, it just seems to be invalidated several times a day.

**Comment by daniellee** [comment]
Any errors messages in the Grafana server log? Any database errors?

(There is a long thread here on various issues around authentication but not sure if any of these have the same issue as you: https://github.com/grafana/grafana/issues/10727)

**Comment by stuartnelson3** [comment]
What jumps out (and was present in the issue you linked) are:
```
2019-04-11T07:36:03.005293501   stdout  t=2019-04-11T07:36:03+0000 lvl=eror msg="Failed to update last_seen_at" logger=context userId=120 orgId=1 uname=<uname> error="Error 1213: Deadlock found when trying to get lock; try restarting transaction"
2019-04-11T07:43:35.143859089   stdout  t=2019-04-11T07:43:35+0000 lvl=eror msg="failed to rotate token" logger=context error="Error 1213: Deadlock found when trying to get lock; try restarting transaction"
2019-04-11T07:43:35.144222955   stdout  t=2019-04-11T07:43:35+0000 lvl=eror msg="Failed to update last_seen_at" logger=context userId=120 orgId=1 uname=<uname> error="Error 1213: Deadlock found when trying to get lock; try restarting transaction"
2019-04-11T07:43:35.168957299   stdout  t=2019-04-11T07:43:35+0000 lvl=eror msg="Failed to update last_seen_at" logger=context userId=120 orgId=1 uname=<uname> error="Error 1213: Deadlock found when trying to get lock; try restarting transaction"
2019-04-11T08:03:48.233097818   stdout  t=2019-04-11T08:03:48+0000 lvl=eror msg="failed to look up user based on cookie" logger=context error="Error 1317: Query execution was interrupted"
2019-04-11T08:03:48.233474715   stdout  t=2019-04-11T08:03:48+0000 lvl=eror msg="failed to look up user based on cookie" logger=context error="Error 1213: Deadlock found when trying to get lock; try restarting transaction"
2019-04-11T08:03:48.235302802   stdout  t=2019-04-11T08:03:48+0000 lvl=eror msg="Unable to load datasource meta data" logger=context userId=0 orgId=1 uname= error="Error 1213: Deadlock found when trying to get lock; try restarting transaction"
2019-04-11T08:26:24.490521201   stdout  t=2019-04-11T08:26:24+0000 lvl=eror msg="Failed to update last_seen_at" logger=context userId=443 orgId=1 uname=<uname2> error="Error 1317: Query execution was interrupted"
```

The solution in the linked issue was talking about using sqlite. We're running on mysql, and the server hasn't seen any changes since the upgrade. It's usage is extremely low. As far as I can tell, this issue only occurs since the upgrade from v5 to v6.

Is there anything I should look in to further on my side?

**Comment by stuartnelson3** [comment]
ping, any status on this?

**Comment by marefr** [comment]
This may be solved by #16785 which will be included in v6.2 stable.

**Comment by rtreffer** [comment]
We (@stuartnelson3 and myself) traced our lock issues down to XtraDB Cluster / Galeras opportunistic locking. We managed to avoid these errors by using just a single db server.
But this should not cause constant signouts, adding more retries sounds good.

**Comment by bergquist** [comment]
Closing this since it should be solved by #16785 ping me if it's not.
